[test_sanitize.cpp:485] | (Runtime) Unit Tests | runtime-unit-tests / runtime_sd_debug_tools_device_dispatch [wh_n300_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [test_sanitize.cpp:485] Value of: error.find("Aborting wait due to watcher… — [CI] MeshWatcherFixture.TensixTestWatcherSanitizeNOCAlignmentL1Write fails: MMIO per-op timeout instead of watcher error
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** (Runtime) Unit Tests
**Job:** runtime-unit-tests / runtime_sd_debug_tools_device_dispatch [wh_n300_civ2]

### Regression timeline

- **Last passing:** commit `1769bd090998` on 2026-07-12T05:08:47Z — https://github.com/tenstorrent/tt-metal/actions/runs/29180740435
- **First failing:** commit `52c1d0e55474` on 2026-07-15T05:04:22Z — https://github.com/tenstorrent/tt-metal/actions/runs/29390511828/job/87483135291
- **Most recent failing:** commit `d424d9a59410` on 2026-07-16T05:06:58Z — https://github.com/tenstorrent/tt-metal/actions/runs/29473041246/job/87565827077

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/29473041246/job/87565827077
- https://github.com/tenstorrent/tt-metal/actions/runs/29390511828/job/87483135291

### Failing test path(s)

- `tests/tt_metal/tt_metal/debug_tools/watcher/test_sanitize.cpp` — test `MeshWatcherFixture.TensixTestWatcherSanitizeNOCAlignmentL1Write` (line 485)
- Binary: `./build/test/tt_metal/unit_tests_debug_tools`

### Error excerpt (terminal failure from logs)

**Run 1 (commit d424d9a59410):**
```
/work/tests/tt_metal/tt_metal/debug_tools/watcher/test_sanitize.cpp:485: Failure
Value of: error.find("Aborting wait due to watcher error") != std::string::npos
  Actual: false
  Expected: true
MMIO per-op timeout: 4B load took 201492 us (budget=2 ms), 4 of 4 bytes remaining.
Location: /work/tt_metal/third_party/umd/device/pcie/device_memcpy.cpp:38
```

**Run 2 (commit 52c1d0e55474):**
```
/work/tests/tt_metal/tt_metal/debug_tools/watcher/test_sanitize.cpp:485: Failure
Value of: error.find("Aborting wait due to watcher error") != std::string::npos
  Actual: false
  Expected: true
MMIO per-op timeout: 4B load took 214232 us (budget=2 ms), 4 of 4 bytes remaining.
Location: /work/tt_metal/third_party/umd/device/pcie/device_memcpy.cpp:38
```

After this GTest assertion failure, the test runner hangs and the job times out after 17 minutes with:
```
##[error]The action 'runtime_sd_debug_tools_device_dispatch [wh_n300_civ2]' has timed out after 17 minutes.
```

### Reproduction steps

On a wh_n300 machine, from the repo root:

```bash
TT_METAL_SLOW_DISPATCH_MODE=1 ./build/test/tt_metal/unit_tests_debug_tools \
  --gtest_filter='MeshWatcherFixture.TensixTestWatcherSanitizeNOCAlignmentL1Write'
```

### Notes

The test `TensixTestWatcherSanitizeNOCAlignmentL1Write` deliberately triggers a NOC write to an unknown/out-of-bounds core (`Unknown core w/ virtual coords 26-18`) and expects the watcher to detect the NOC error and throw an exception containing `"Aborting wait due to watcher error"`. Instead, it catches an MMIO per-op timeout from `tt::umd::memcpy_from_device` (via `device_memcpy.cpp:38`) — meaning the watcher did not fire/propagate the error fast enough (or at all) before the MMIO timeout was hit. The call stack terminates in `tt::Cluster::dram_barrier` → `LaunchProgram` → `SDMeshCommandQueue::dispatch_program`.

The subsequent hang (job-level 17-minute timeout) is a consequence of the device being left in a bad state after the test failure.

**Determinism check:**
- Run 1 terminal error: `test_sanitize.cpp:485: Failure` — `MMIO per-op timeout: 4B load took 201492 us (budget=2 ms)`
- Run 2 terminal error: `test_sanitize.cpp:485: Failure` — `MMIO per-op timeout: 4B load took 214232 us (budget=2 ms)`
- Both: same file, same line, same assertion (`error.find("Aborting wait due to watcher error")` → false), same call stack, numeric timing difference only.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (Runtime) Unit Tests`
`Auto-triage-job-name: runtime-unit-tests / runtime_sd_debug_tools_device_dispatch [wh_n300_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

--- a/tests/tt_metal/tt_metal/debug_tools/watcher/test_sanitize.cpp
+++ b/tests/tt_metal/tt_metal/debug_tools/watcher/test_sanitize.cpp
@@ -238,6 +238,11 @@ void RunTestOnCore(
         }
         // (gen1 path: no CTA bindings needed; the kernel runs on exactly one DM processor.)
 
+        // Under SD the kernel signals completion via RUN_MSG_DONE, not the FD notify path (which wedges the NOC).
+        if (fixture->IsSlowDispatch()) {
+            defines["WATCHER_KERNEL_SLOW_DISPATCH"] = "1";
+        }
+
         // Select the DM config variant matching the current architecture (Gen2 on Quasar, Gen1 otherwise).
         auto gen1_processor =
             use_ncrisc ? tt::tt_metal::DataMovementProcessor::RISCV_1 : tt::tt_metal::DataMovementProcessor::RISCV_0;

--- a/tests/tt_metal/tt_metal/test_kernels/dataflow/dram_copy_to_noc_coord_2_0.cpp
+++ b/tests/tt_metal/tt_metal/test_kernels/dataflow/dram_copy_to_noc_coord_2_0.cpp
@@ -77,8 +77,9 @@ void kernel_main() {
     // Dispatcher Kernel is able to finish.
     // Device::close() requires fast-dispatch kernels to finish.
     volatile tt_l1_ptr go_msg_t* go_message_in = GET_MAILBOX_ADDRESS_DEV(go_messages[0]);
-#if defined(COMPILE_FOR_DM)
-    // SD signaling: Quasar DM requires RUN_MSG_DONE. TODO: remove once FD is enabled on Quasar.
+    // SD enabled on all archs: notify completion via RUN_MSG_DONE to mailbox. FD notify path
+    // posts to a dispatcher absent under SD and wedges the NOC.
+#if defined(WATCHER_KERNEL_SLOW_DISPATCH)
     go_message_in->signal = RUN_MSG_DONE;
 #else
     uint64_t dispatch_addr = calculate_dispatch_addr(go_message_in);

